src/uct/tcp/tcp_ep.c

```diff
@@ -62,64 +61,203 @@ static void uct_tcp_ep_epoll_ctl(uct_tcp_ep_t *ep, int op)
 
 static inline int uct_tcp_ep_can_send(uct_tcp_ep_t *ep)
 {
-    ucs_assert(ep->offset <= ep->length);
+    ucs_assert(ep->tx.offset <= ep->tx.length);
     /* TODO optimize to allow partial sends/message coalescing */
-    return ep->length == 0;
+    return ep->tx.length == 0;
+}
+
+static inline void uct_tcp_ep_ctx_rewind(uct_tcp_ep_ctx_t *ctx)
+{
+    ctx->offset = 0;
+    ctx->length = 0;
+}
+
+static ucs_status_t uct_tcp_ep_ctx_init(uct_tcp_iface_t *iface,
+                                        uct_tcp_ep_ctx_t *ctx)
+{
+    ucs_assertv(ctx->buf == NULL, "ctx=%p", ctx);
+
+    ctx->buf = ucs_malloc(ucs_max(iface->config.buf_size,
+                                  iface->config.short_size), "tcp_buf");
+    if (ctx->buf == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    uct_tcp_ep_ctx_rewind(ctx);
+
+    return UCS_OK;
+}
+
+static void uct_tcp_ep_ctx_cleanup(uct_tcp_ep_ctx_t *ctx)
+{
+    if (ctx->buf != NULL) {
+        ucs_free(ctx->buf);
+        ctx->buf = NULL;
+    }
+}
+
+void uct_tcp_ep_ctx_migrate(uct_tcp_ep_ctx_t *to_ctx, uct_tcp_ep_ctx_t *from_ctx)
+{
+    ucs_assertv(to_ctx->buf == NULL, "to_ctx=%p", to_ctx);
+    ucs_assertv(from_ctx->buf != NULL, "from_ctx=%p", from_ctx);
+
+    *to_ctx = *from_ctx;
+
+    memset(from_ctx, 0, sizeof(*from_ctx));
+}
+
+static void uct_tcp_ep_addr_cleanup(ucs_sock_addr_t *sock_addr)
+{
+    if (sock_addr->addr != NULL) {
+        ucs_free((void*)sock_addr->addr);
+    }
+
+    memset(sock_addr, 0, sizeof(*sock_addr));
+}
+
+static ucs_status_t uct_tcp_ep_addr_init(ucs_sock_addr_t *sock_addr,
+                                         const struct sockaddr *addr)
+{
+    socklen_t addr_len = sizeof(struct sockaddr_in); /* TODO */
+    struct sockaddr *new_addr;
+
+    if (addr == NULL) {
+        memset(sock_addr, 0, sizeof(*sock_addr));
```

Comment:
better set each field separately, since sock_addr is not a buffer
e.g sa->addr = NULL; sa->addrlen = 0